server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java

```diff
@@ -138,6 +138,123 @@ public void testParseWithShadowedMultiField() throws Exception {
         assertNotNull(doc.rootDoc().getField("field.keyword"));
     }
 
+    public void testRuntimeFieldAndArrayChildren() throws IOException {
+        DocumentMapper mapper = createDocumentMapper(topMapping(b -> {
+            b.field("dynamic", "true");
+            b.startObject("runtime");
+            {
+                b.startObject("object").field("type", "test").endObject();
+            }
+            b.endObject();
+        }));
+
+        {
+            ParsedDocument doc = mapper.parse(source(b -> {
+                b.startObject("object");
+                b.array("array", 1, 2, 3);
+                b.field("foo", "bar");
+                b.endObject();
+            }));
+            assertNotNull(doc.rootDoc().getField("object.foo"));
+            assertNotNull(doc.rootDoc().getField("object.array"));
+        }
+
+        {
+            ParsedDocument doc = mapper.parse(source(b -> {
+                b.startArray("object");
+                {
+                    b.startObject().array("array", 1, 2, 3).endObject();
+                    b.startObject().field("foo", "bar").endObject();
+                }
+                b.endArray();
+            }));
+            assertNotNull(doc.rootDoc().getField("object.foo"));
+            assertNotNull(doc.rootDoc().getField("object.array"));
+        }
+    }
+
+    public void testRuntimeFieldDoesNotShadowObjectChildren() throws IOException {
+        DocumentMapper mapper = createDocumentMapper(topMapping(b -> {
+            b.field("dynamic", "true");
```

Comment:
nit: maybe DynamicMappingsTests#dynamicMapping should be moved to MapperServiceTestCase? Also, isn't dynamic true the default anyways?

Reply:
`dynamic:true` is indeed the default but I wanted to make it clear in the test.  Nice idea about moving `dynamicMapping`, I'll pull that in.

Reply:
Although it actually doesn't help in this case because it always uses `properties` and we want to add things to `runtime` here.  Meh.